### ISSUE TITLE
Issues with creating a fingerprint / hashing the certificate

### ISSUE THREAD

**Issue by d-Rickyy-b** [google/certificate-transparency-go#947]
Hi there, I am currently trying to calculate the hashes/fingerprints of certificates.
From my understanding the fingerprint of a certificate is the hashsum of the raw DER encoded bytes. That data should be contained in the Certificate.Raw field. I am using the following snippet.

```go
import (
    "crypto/sha256"
    "github.com/google/certificate-transparency-go/x509"
)

[...]

// certificate is an instance of x509.Certificate
h := sha256.New()
h.Write(certificate.Raw)
hashSum := h.Sum(nil)
hashstr := fmt.Sprintf("%02x", hashSum)
fmt.Println(hashstr)
```

Now the certificate I am working on was randomly selected and downloaded via this library from the transparency log -> https://crt.sh/?id=6919340245.

Its SHA2 hash should be `01B6248E21382C82471660BC3427EBE12808C46DD022FC2D15DCB940FE9DF235`.
But using my code I get `810720ED813E9C0E63F04D53F0F8E99DF4092BB82A0512DCFEFACA35B3B2CCC3`.

Same thing for SHA1. It should be: `955F8A0D03611A4BD21CB64183B6720C6EB181EC`
I get: `BADC8B59485564F3CA6328CDCF59CF32957D95BF`

The Write method also tells me that 802 bytes have been processed - the exact same amount of bytes held in certificate.Raw.

EDIT: I also tried the certificate.RawTBSCertificate field, but it contains the exact same data as the Raw field.
![grafik](https://user-images.githubusercontent.com/5798157/173256197-42da9abd-47ce-4fc6-be3a-7544294a5063.png)

I really don't understand why I get wrong hashes. Maybe you are able to detect an issue with my code or can explain what I am doing wrong. Thank you very much in advance.

Cheers,
Rico

**Comment by mhutchinson** [comment]
Hi Rico, this question may be better directed to crt.sh if you are trying to match their calculation of a fingerprint. At a quick scan, it looks like it is calculated here: https://github.com/crtsh/ct_monitor/blob/master/ct_monitor.go#L660

From a crypto proof perspective (i.e. the spec defined by rfc6962), the leaf hash is calculated using the raw certificate and prepending a constant `0` before it, as seen here:
https://github.com/transparency-dev/merkle/blob/main/rfc6962/rfc6962.go#L49

As an example of using this library to verify that this certificate is committed to by a log checkpoint/STH, see:
https://github.com/google/trillian-examples/blob/master/clone/cmd/ctverify/ctverify.go#L60-L68

**Comment by pphaneuf** [comment]
Note: the leaf hash is *not* calculated using the raw certificate, it's calculating using the leaf input data, and is a completely different type of thing (there can be multiple leaf hashes for the same raw certificate).

**Comment by d-Rickyy-b** [comment]
Thanks for your answers! Special thanks to @AlCutter, you made me rethink my approach.

I found that using the `ct.RawLogEntry.Cert.Data` as input works just fine for calculating the hash/fingerprint of the certificate, because that seem to be the actual DER bytes as a whole.

My issue was that I transformed the RawLogEntry by calling `.ToLogEntry()` on it and using `logEntry.X509Cert.Raw` as input for the hashing algorithm. Sadly the "Raw" cert does not contain the same data (full DER bytes) as the `ct.RawLogEntry.Cert.Data`.

Maybe someone of you knows what's the difference between the `ct.RawLogEntry.Cert.Data` and the `logEntry.x509Cert.Raw`.
